development_guidelines.md

```diff
@@ -0,0 +1,415 @@
+    - [How-to change logging level for a package](#how-to-change-logging-level-for-a-package)
+  - [How-to write proper tests](#how-to-write-proper-tests)
+  - [How-to deal with threads in tests](#how-to-deal-with-threads-in-tests)
+  - [Gtest command line flags](#gtest-command-line-flags)
+  - [How-to run particular test of test suite](#how-to-run-particular-test-of-test-suite)
+- [Mocking the hardware](#mocking-the-hardware)
+- [FAQ](#faq)
+
+## Default development branch
+Our current default development branch is melodic-devel 
+(see [here](https://github.com/PilzDE/pilz_robots/branches)).
+All new feature are developed for 
+[ROS Melodic Morenia](https://www.ros.org/reps/rep-0003.html#melodic-morenia-may-2018-may-2023) 
+which primarily targets Ubuntu Bionic (18.04) (for more information see 
+[here](http://wiki.ros.org/melodic)).  
+For backwards compatibility reasons we still support 
+[ROS Kinetic Kame](https://www.ros.org/reps/rep-0003.html#kinetic-kame-may-2016-may-2021)
+in combination with Ubuntu Xenial (16.04).
+In order to support ROS Kinetic Kame, we proceed as follows:
+- First, we develop the new feature for ROS Melodic Morenia and merge the
+new feature into the melodic-devel branch.
+- Secondly, we cherry-pick the new feature from the melodic-devel branch and
+merge it into the kinetic-devel branch.
+
+
+## Review process
+Before any pull request is approved a developer not involved in 
+the development of the feature has to review the pull request.  
+  
+Our goal is to automate the review process as much as possible by including
+as many checks as possible into the continuous integration process.
+However, some things are currently manual or simply cannot be automated.  
+In general the reviewer has to check/review the following:
+- Specification:
+  - The feature and its changes to the system are comprehensively specified.
+  - All requirements are exhaustively tested by one or more tests.
+  - All specifications are linked against their corresponding tests 
+(how-to link see [here](#how-to-link-a-requirement-against-a-test))
+- User interface:
+  - The user interface is easy to understand and use.
+  - The user interface is comprehensively documented.
+  - If an existing user interface is broken by the new feature, then the
+interface version has to be adapted according the nature of the change.
+- System architecture:
+  - The system architecture complies with the clean architecture guidelines.
+(see "Clean Architecture" by Robert C. Martin)
+  - The architecture diagrams are up to date and easily understandable.
+- Code:
+  - The code has to comply with the clean code rules 
+(see "Clean Code: A Handbook of Agile Software Craftsmanship" by Robert C. Martin).
+- Tests:
+  - The tests are easy to understand (e.g. each test only checks one aspect 
+of a component, etc.)
+  - The tests are comprehensively documented (how-to document tests see
+[here](#how-to-document-tests))
+
```

Comment:
- Documentation
  - The creation of the documentation via rosdoc_lite shows no errors

Reply:
Currently, rosdoc_lite has errors not related to us. Until these errors are fixed, I'm against this statement.

Reply:
- Documentation
  - The creation of the documentation via rosdoc_lite shows no errors related to your code

?

Reply:
Remains left out -> Open new PR later on. -> Done.